docs/dev/data-identity.rst

```diff
@@ -0,0 +1,148 @@
+predicting the lifetime of objects in the Lua environment can be difficult.
+If it is necessary to create a ``type_identity`` that will not have program lifetime,
+it is incumbent on the developer to ensure that no references of that type identity persist beyond its lifetime.
+
+Due to the way template types are implemented in the C++ compilers we use for Dwarf Fortress, any specialization of one
+of the type identity classes noted above as a template must at present be statically constructed in the DFHack core.
+This is because we export the statically constructed instances from the core library to plugins, which then imports them
+from the core library instead of instantiating them locally. As a result, referencing an instance in a plugin that has
+not been instantiated in the core will result in linkage errors when linking the plugin to the core library.
+
+We could instead _not_ export the templated types, and thus their statically constructed identity objects,
+and instead allow the compiler to instantiate a local copy of these instances while compiling a plugin,
+but this would definitely result in a violation of the current requirement that there be at most one instance of the
+type identity object for a given C++ type across the entire program (including plugins loaded as a shared library).
+
+For primitive and opaque types the static constructors of the identity types
+are generally found in ``DataIdentity.h`` or ``DataIdentity.cpp``.
+Types defined by Dwarf Fortress are constructed in the header files and the related ``static*.inc`` files created by codegen,
+which are included into DFHack via ``DataStatics.cpp``.
+
+Some plugins also define their own type identities. Type identities in plugins should be used with caution,
+because the DFHack plugin model allows plugins to be unloaded on request.
+Since the ``type_identity`` object is constructed within the the plugin's address space, and Lua objects that reference
+this ``type_identity`` object will hold a (borrowed) pointer to that object,
+unloading the plugin will result in a dangling pointer reference within the Lua environment.
+It is, at present, incumbent on plugin authors to ensure that they do not use plugin defined type identities on objects
+that will persist in the Lua environment beyond the lifetime of the plugin.
+Declaring a ``struct_identity`` in a plugin that is the child of another ``struct_identity`` will also result in
+a potentially dangling reference to that identity in the ``child`` vector of the parent identity, which means this
+must also be approached with caution.
+
+Namespace issues
+================
+
+The type identity system formally lives in the ``DFHack`` namespace.
+However, because the types created by the codegen process live in the ``df`` namespace,
+the identities needed to describe types coming from Dwarf Fortress are imported into the ``df`` namespace
+and at least some types are _only_ defined in that namespace.
```

Comment:
could you give/link to an example here?

Reply:
the actual issue is that while the `type_identity` objects are in the `DFHack` namespace, the `identity_traits` trait class is in the `df` namespace

so i am actually mistaken there, and will rewrite this section